Create index.eng.md (#28)
planetarium/snack.planetarium.dev: content/posts/2019/06/xunit-unity-runner/index.eng.md

---
title: Unit Testing in Unity Environment
date: 2019-06-28
authors: [hong.minhee]
translators: [kidon.seo]

---

> Hello, I'm Hong Minhee, [Libplanet] commiter at Planetarium. 

Reviewer: ```suggestion
Hello, I'm Hong Minhee, [Libplanet] committer at Planetarium. 
```

> In this post, I want to talk about why we came to a conclusion to conduct Unit Tests on Unity, 

Reviewer 2: 단위 테스트를 실행한다는 게 1회 시행한다는 것은 아니고, 매 수정마다 자동으로 돌아가는 테스트를 Unity에서도 돌리겠다, 이런 뜻이라서 “conduct”가 맞을지 모르겠네요.

```suggestion
In this post, I want to talk about why we came to a conclusion to run unit tests on Unity too, 
```

Author: 네 저도 그런 맥락에서 번역을 했는데 이렇게 수정해도 좋을 것 같아요. 한 가지 궁금한게 있는데 유니티 말고도 유닛테스트를 어디서 돌리나요? 이 부분의 뉘앙스를 맞추는데 도움이 될거 같아서요!

Reviewer 2: 저희가 [continuous integration](https://en.wikipedia.org/wiki/Continuous_integration)이라고 해서 커밋을 할 때마다 혹시 그 변경에 의해 기존에 동작하던 게 오작동하게 되지는 않는지 (regression) 등을 자동으로 테스트하는데요, 이 때 Linux, macOS, Windows 및 여러 가지 .NET 플랫폼에서 테스트가 돌아가고 있습니다. 이 글은 이 조합에 Unity 환경도 더해졌다는 이야기였습니다.

Author: 네 그럼 이 문장 전체를 "In this post, I want to talk about our decision to run unit tests in the widely used Unity environment as well, along with our journey in approaching this task."로 바꾸는게 문장 의도에 더 맞을 것 같습니다. 제가 처음에 올린 번역문장은 뭔가 "널리 쓰이는 Unity 환경"을 강조하는 느낌이어서 살짝 이상했는데 홍님 피드백을 들어보니 저 위에 문장이 더 적합해 보입니다. (근데 제 글을 제가 어떻게 수정하는지 몰라서... 수정 부탁드리겠습니다ㅠ)

the most widely used game engine, and how we actually approached it.

Supporting Different Environments on Libplanet
---------------------------------------

Libplanet is a common library that solves game implementation problems such as P2P communication and 

> data synchronization when creating online multiplay games that run on distributed P2P.

Reviewer 2: ```suggestion
data synchronization when creating online multiplayer games that run on distributed P2P.
```

Author: "Multiplayer"라는 단어가 더 적합한 것 같습니다. 근데 이 번역은 기존 Libplanet 소개글에서 계속 copy&paste한 부분이어서 이전 포스트들도 이렇게 수정해야할 거 같네요?

Reviewer 2: 그럼 이전 글들은 제가 고쳐놓겠습니다.

> Testing codes, especially unit tests, are needed to achieve rapid improvement while minimizing malfunctions 

Reviewer 2: ```suggestion
Automated tests, especially unit tests, are needed to achieve rapid improvement while minimizing malfunctions 
```

that are prone to [regressions] or corner cases. Furthermore, Libplanet is a library and because it is 
difficult to determine which operating system and .NET runtime each game or app will use, we need to run all 
tests in as many different environments as possible.

> So our team ran tests on [Azure Pipelines][^1] with (Lux, macOS, Windows) × (.NET Framework, Mono, .NET Core) 

Reviewer 2: ```suggestion
So our team had run tests on [Azure Pipelines][^1] with (Linux, macOS, Windows) × (.NET Framework, Mono, .NET Core) 
```

> combination[^2] whenever a push or a pull was requested in the Libplanet repository.

Reviewer 2: ```suggestion
combination[^2] whenever a push or a pull request was made in the Libplanet repository.
```

{{< 
figure 
src="before.png" 
width="298" 
caption="The combination of environments tested on each build"
>}}

Unity ≠ Mono
------------

At first, we thought this was enough because Unity uses Mono Runtime. But as we used Libplanet to develop 

> our game on Unity, we had to watch unexpected moves in the game several times, and it became increasingly 

Reviewer 2: ```suggestion
our game on Unity, we had to encounter unexpected behaviors in the game several times, and it became increasingly 
```

evident that just passing the test on Mono was not enough.

[In fact, the Mono used in Unity seems to be a fairly long-standing downstream, with a lot of patches added 
to the upstream.][unity-mono] And even if they were tested at the exact same Mono runtime, there were a lot 

> of special conditions created by Unity players. For instance, [NetMQ], [ZeroMQ]'s C# implementation, had 

Reviewer 2: ```suggestion
of special conditions created by Unity Player. For instance, [NetMQ], [ZeroMQ]'s C# implementation, had 
```

numerous library malfunctions due to many complicated things happening inside compared to the simplicity of 
the APIs revealed on the outer layer.

All of these considerations led to an agreement that a testing environment for Unity needed to be added to CI 
to ensure reliable functionality.

Testing xUnit.net on Unity
--------------------------

Because there is a unit testing feature available in Unity, we tried to use it at first. Unfortunately, Unity's 

> unit testing was done in an editor environment used by game developers, not in a player environment, and the 

Reviewer 2: ```suggestion
unit testing was done in an Editor environment used by game developers, not in a Player environment, and the 
```

testing framework supported only [NUnit]. We thought about changing all of Libplanet's [xUnit.net]-based test 
codes, to NUnit, but with such a high volume of codes to change at once, we didn't want to risk making mistakes 
that are hard to notice.

So we decided to create a test runner app instead of a game app with Unity. Fortunately, xUnit.net was well divided 

> between APIs for writing tests and APIs for running tests. This is probably due to the diverse front end plug-ins to 

Reviewer 2: ```suggestion
between APIs for writing tests and APIs for running tests. This is probably due to the diverse frontend plug-ins to 
```

various IDE, GUI and CLI. In fact, if you search "xunit runner" on NuGet, you'll get a xUnit.net test runner for a 
variety of environments.

A downside is that because there are no API documents, we had to search the source code of xUnit.net and the source 
code of other test runners.[^3]

The test runner API on xUnit.net is roughly as follows: First, the client code looks for test classes and the tests 
methods within those classes from the assembly files (*.dll*) passed through the input. After, the client code can 
decide which test to run. Then, the test cases are run by the test runner. Because the test discovery and execution 

> can be done in parallel for performance, the API follows a typical [<abbr title= “inversion of control”>IoC</abbr>[IoC] 

Reviewer 2: ```suggestion
can be done in parallel for performance, the API follows a typical [<abbr title="inversion of control">IoC</abbr>][IoC] 
```

> pattern. An interface called [‘IMessageSinkWithTypes’][IMessageSinkWithTypes], which receives events such as test 

Reviewer 2: ```suggestion
pattern. An interface called [`IMessageSinkWithTypes`][IMessageSinkWithTypes], which receives events such as test 
```

discovery, start running, failure, success, skip, and so on in the form of a message, must be implemented in the client 
code to show the log on screen when such events occur. Because our team didn't run the tests in parallel, it was quite 
frustrating to have a less liberal API with a lengthy client code. 🙄

> Building a <abbr title= “command-line interface”>CLI</abbr> with Unity

Reviewer 2: ```suggestion
Building a <abbr title="command-line interface">CLI</abbr> with Unity
```

----------------------------------------------------------------------

Our greatest concern when creating a Unity test runner was that since this was to be done on 

> <abbr title= “continuous integration”>CI</abbr> in the first place, we believed that the test runner had to be manipulated 

Reviewer 2: ```suggestion
<abbr title="continuous integration">CI</abbr> in the first place, we believed that the test runner had to be manipulated 
```

> by the <abbr title= “command-line interface”>CLI</abbr> rather than by the graphics screen, and the results should be 

Reviewer 2: ```suggestion
by the <abbr title="command-line interface">CLI</abbr> rather than by the graphics screen, and the results should be 
```

visible. So with Unity being a platform for making graphic games, we were worried whether it'd be a good idea or not to 
actually create a CLI app.

Fortunately, we found out that Unity has a <dfn>headless mode</dfn>, which means that all logs taken in 

> [‘Debug.Log()’][Debug.Log()] method are output as [standard output][stdout], without graphical display.

Reviewer 2: ```suggestion
[`Debug.Log()`][Debug.Log()] method are output as [standard output][stdout], without graphical display.
```

> [`Debug.Log()`][Debug.Log()] method are output as [standard output][stdout], without graphical display.

Reviewer: ```suggestion
[`Debug.Log()`][Debug.Log] method are output as [standard output][stdout], without graphical display.
```

{{< 
figure 
src="unity-build-settings.png" 
width="356" 
caption="<q>Server Build</q> option in Unity build settings that turn on headless mode"
>}}

> Even if you don't use the ‘Debug.Log()’ method provided by Unity, we've also figured out that just like creating a 

Reviewer 2: ```suggestion
Even if you don't use the `Debug.Log()` method provided by Unity, we've also figured out that just like creating a 
```

> typical application the [‘Console’][Console] class provided by the .NET standard works as well.

Reviewer 2: ```suggestion
typical application the [`Console`][Console] class provided by the .NET standard works as well.
```

> However, since the ‘Main()’ method cannot be defined, the command line factor was not accepted as the ‘string[] args’ 

Reviewer 2: ```suggestion
However, since the `Main()` method cannot be defined, the command line factor was not accepted as the `string[] args`
```

> factor of the ‘Main()’ method, but instead had to be obtained as the [‘Environment.GetCommandLineArgs()’][GetCommandLineArgs] 

Reviewer 2: ```suggestion
parameter of the `Main()` method, but instead had to be obtained as the [`Environment.GetCommandLineArgs()`][GetCommandLineArgs] 
```

> method. Similarly, the program's termination required an explicit call to the [‘Application.Quit()’][Quit] method to 

Reviewer 2: ```suggestion
method. Similarly, the program's termination required an explicit call to the [`Application.Quit()`][Quit] method to 
```

terminate the process directly.

Lastly, there were messages being output from Unity player itself, but we couldn't find a way to block it, so we had 
to wrap it up.[^4]

{{< 
figure 
src="noisy-output.png" 
width="739" 
caption="Unity player's own message, printed on the first and last line, was never removed."
>}}

Build Automation
----------------

> Building a CLI app with Unity and writing a document on how to build one on Windows as well as Linux or MacOS makes 

Reviewer 2: ```suggestion
Building a CLI app with Unity and writing a document on how to build one on Windows as well as Linux or macOS makes 
```

the process tricky and easier for people to get inconsistent results. So we decided to make a tag in the repository 

> and when pushed, and it'll automatically build for Linux, MacOS, and Windows.

Reviewer 2: ```suggestion
and when pushed, and it'll automatically build for Linux, macOS, and Windows.
```

Although we thought about putting CI on the board, we decided that it was unnecessary and used [GitHub Actions] to build it.

> [Referring to Kawai Yoshifumi(河合宜文)'s post][5], we were able to carry out the entire build process inside the Docker. 

Reviewer 2: ```suggestion
[Referring to Kawai Yoshifumi's post][5], we were able to carry out the entire build process inside the Docker. 
```

In the process, we experienced things that we hadn't experienced in other environments:

Because Unity was a commercial product, we needed to activate the license.

Unity has somewhat an ambiguous boundary between an editor and a player. The code that will run in the editor environment 
can also be scripted, and this enables the app to be built by itself because Unity includes the app-building script as 
part of the app and then runs it. 

At first, we thought we needed to build on all three operating systems, but fortunately, Unity supported [cross-compiling][6] 

> and we were able to build the app for MacOS and Windows on Linux.

Reviewer: ```suggestion
and we were able to build the app for macOS and Windows on Linux.
```

[Travis CI]: https://travis-ci.com/
[unity-mono]: https://github.com/Unity-Technologies/mono
[ZeroMQ]: http://zeromq.org/
[NetMQ]: https://github.com/zeromq/netmq
[NUnit]: https://nunit.org/
[xUnit.net]: https://xunit.net/
[xmldoc]: https://docs.microsoft.com/dotnet/csharp/programming-guide/xmldoc/
[IoC]: https://en.wikipedia.org/wiki/Inversion_of_control
[IMessageSinkWithTypes]: https://github.com/xunit/xunit/blob/2.4.1/src/xunit.runner.utility/Messages/IMessageSinkWithTypes.cs
[Debug.Log]: https://docs.unity3d.com/ScriptReference/Debug.Log.html
[stdout]: https://en.wikipedia.org/wiki/Standard_streams#Standard_output_(stdout)
[Console]: https://docs.microsoft.com/dotnet/api/system.console
[GetCommandLineArgs]: https://docs.microsoft.com/dotnet/api/system.environment.getcommandlineargs
[Quit]: https://docs.unity3d.com/ScriptReference/Application.Quit.html
[GitHub Actions]: https://github.com/features/actions
[5]: https://medium.com/@neuecc/using-circle-ci-to-build-test-make-unitypackage-on-unity-9f9fa2b3adfd
[6]: https://en.wikipedia.org/wiki/Cross_compiler
[7]: https://github.com/planetarium/xunit-unity-runner
[8]: https://github.com/planetarium/xunit-unity-runner/releases

> [^1]: As of June 2019, <abbr title= “continuous integration”>CI</abbr> services that support all Linux, MacOS, and 

Reviewer: The footnote should be written on one line so it will display properly.
```suggestion
[^1]: As of June 2019, <abbr title= “continuous integration”>CI</abbr> services that support all Linux, macOS, and Windows include [Travis CI] and Azure Pipelines. Our team used Travis CI at first, but it didn't perform well, so we're now using Azure Pipelines. 
```

> Windows include [Travis CI] and Azure Pipelines. Our team used Travis CI at first, but it didn't perform well, so 

Reviewer: ```suggestion
```

> we're now using Azure Pipelines. 

Reviewer: ```suggestion
```

[^2]: Because the .NET Framework supports only Windows, it will be tested in 7 environments instead of 9. 

> [^3]: Because the .NET <abbr title="integrated development environment">IDE</abbr> has become very common for quite 

Reviewer: ```suggestion
[^3]: Because the .NET <abbr title="integrated development environment">IDE</abbr> has become very common for quite some time, there are many projects that don't post API documents on the Web and simply leave [XML document annotations][xmldoc] in the source code. Those annotations will appear small as a tooltip when the class or method is automatically completed in IDE.
```

> some time, there are many projects that don't post API documents on the Web and simply leave [XML document annotations][xmldoc] 

Reviewer: ```suggestion
```

> in the source code. Those annotations will appear small as a tool tip when the class or method is automatically completed in IDE. 

Reviewer: ```suggestion
```
